### PR TITLE
Implementing HTTP Server-Sent Events

### PR DESCRIPTION
**Pull request by akvasha**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Implementing HTTP Server-Sent Events

Detailed description / Documentation draft:

Server-Sent Events implementation, the first draft commit
The implementation is being actively pursued.

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by filimonov** [comment]
Looks like lot of copied code there. Is it just a proof-of-concept or there are some fundamental problems which don't allow to reuse existing code (for example by inheritance)?

**Comment by akvasha** [comment]
> Looks like lot of copied code there. Is it just a proof-of-concept or there are some fundamental issues which don't allow to reuse existing code (for example by inheritance)?

That's just an init commit, which is currently being actively pursued. Mostly copied code here is used to keep based HTTP handler correct which I use for debugging purposes. Of course, this code will be refactored and part of it already differs, however some code may still be duplicated as discussed with @ alexey-milovidov.

So, basically, we could say that it was just proof-of-concept.

File: src/Server/HTTPStreamHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,1100 @@
...
+    encoder << decoded;
+    encoder.close();
+    return ostr.str();
+}
+
+static Poco::Net::HTTPResponse::HTTPStatus exceptionCodeToHTTPStatus(int exception_code)
+{
+    using namespace Poco::Net;
+
+    if (exception_code == ErrorCodes::REQUIRED_PASSWORD)
```
Comment (danlark1) [inline src/Server/HTTPStreamHandler.cpp:136]: Why not switch case? Easier to read for primitives

File: src/Server/HTTPStreamHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,1100 @@
...
+    : server(server_)
+    , log(&Poco::Logger::get(name))
+    , param_name(param_name_)
+{
+    server_display_name = server.config().getString("display_name", getFQDNOrHostName());
+}
+
+HTTPStreamHandler::~HTTPStreamHandler()
+{
+    (void)this;
```
Comment (danlark1) [inline src/Server/HTTPStreamHandler.cpp:270]: ?

File: src/Server/HTTPStreamHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,1100 @@
...
+    else
+    {
+        if (!request_credentials)
+            request_credentials = request_context->makeGSSAcceptorContext();
+
+        auto * gss_acceptor_context = dynamic_cast<GSSAcceptorContext *>(request_credentials.get());
+        if (!gss_acceptor_context)
+            throw Exception("Invalid authentication: unexpected 'Negotiate' HTTP Authorization scheme expected", ErrorCodes::AUTHENTICATION_FAILED);
+
+#pragma GCC diagnostic push
```
Comment (danlark1) [inline src/Server/HTTPStreamHandler.cpp:357]: Looks weird, comment? Or maybe it is correct that GSSAcceptorContext is not the child of request_credentials.get()?

**Pull request closed without merging**